External/FEXCore/Source/Interface/Core/ObjectCache/ObjectCacheService.h

```diff
@@ -128,17 +234,101 @@ namespace FEXCore::CodeSerialize {
 
         class WorkItemRemoveNamedRegion : public NamedRegionWorkItem {
           public:
-            WorkItemRemoveNamedRegion(uint64_t base, uint64_t size, CodeRegionMapType::iterator entry)
+            WorkItemRemoveNamedRegion(uint64_t base, uint64_t size, std::unique_ptr<CodeRegionEntry> entry)
               : NamedRegionWorkItem {NamedRegionJobType::JOB_REMOVE_NAMED_REGION}
               , Base {base}
               , Size {size}
-              , Entry {entry} {}
+              , Entry {std::move(entry)} {}
 
             uint64_t Base;
             uint64_t Size;
-            CodeRegionMapType::iterator Entry;
+            std::unique_ptr<CodeRegionEntry> Entry;
         };
       /**  @} */
+
+    private:
+      NamedRegionObjectHandler *NamedRegionHandler;
+      CodeObjectSerializeService *CodeObjectCacheService;
+  };
+
+  class NamedRegionObjectHandler final {
+    public:
+      NamedRegionObjectHandler(FEXCore::Context::Context *ctx);
+
+      void HandleNamedRegionObjectJobs();
+
+      CodeObjectSerializationConfig const &GetDefaultSerializationConfig() const {
+        return DefaultSerializationConfig;
+      }
+
+    protected:
+      friend class AsyncJobHandler;
+
+      // Return a default code header based off the default serialization config
+      CodeObjectSerializationHeader DefaultCodeHeader(uint64_t Base, uint64_t Offset) const {
+        return CodeObjectSerializationHeader {
+          .Config = DefaultSerializationConfig,
+          .OriginalBase = Base,
+          .OriginalOffset = Offset,
+          .NumCodeEntries = 0,
+          .NumRelocationsTo = 0,
+          .TotalRelocationsCount = 0,
+        };
+      }
+
+      /**
+       * @brief Adds an asynchronous add named region work item to the object queue
+       *
+       * This adds the job that will do the loading of file resources and data tracking.
+       */
+      void AsyncAddNamedRegionWorkItem(const std::string &base, const std::string &filename, bool executable, CodeRegionMapType::iterator entry) {
```

Comment:
Why do we need async versions of this?

Reply:
Because we can't block the JIT on IO.

Reply:
sure JIT->MMAP->Named Region.
It's significant IO. Add it to an async job queue to stop blocking JIT execution.

Reply:
I think we should design the system in a way that substantial IO doesn't need to happen on file load.

Performance wise it is probably better, and far easier to reason about, small stutters on code file mmaps vs other complications or unpredictable performance later on.

Reply:
The act of loading the index required to know what is in the file is enough IO to make the JIT slow down.
It's unfeasible to lower IO to the point that it doesn't impact JIT performance.

Reply:
Can't the index be flatted down and then just mmap'd, similar to what AOTIR does?

Reply:
Also WRT issue, it's probably best we keep separate "index" and "object code"/"ir code" files, as the indexes need to be rebuilt while the code files can just be appended to.

Reply:
The entire file is just mmap'd already. flattening won't work since this file is multi-process shared and these caches get incrementally appended from multiple processes.

Reply:
Flattening can work just fine with multi process, why wouldn't it?

Reply:
(can just use a red/black tree that uses local offsets instead of pointers)

Reply:
Well. It's not how I'm implementing it regardless.